docs/development/machine_error_codes.md

```diff
@@ -399,6 +399,59 @@ type GetVolumeIDsResponse struct {
 The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
 This string MAY be surfaced by MCM to end users.
 
+#### `MigrateMachineClass`
+
+A Provider can OPTIONALLY implement this driver call. Else should return a `NOT_IMPLEMENTED` status in error.
+This driver call will be called by the Machine Controller to try to perform a machineClass migration for an unknown machineClass Kind. This helps in migration of one kind of machineClass to another kind. For instance an machineClass custom resource of `AWSMachineClass` to `MachineClass`.
+
+- On successful migration of machine class the Provider MUST reply `0 OK` (or) `nil` error.
+- `MigrateMachineClassRequest` expects the `ClassSpec` containing the `type ClassSpec struct`. It also expects a `MachineClass` client that can be used to communicate with the cluster containing the machine objects. It also passes the `Namespace` where these objects can be found.
+- `MigrateMachineClass` is responsible for creating the new `CR` of the new kind (MachinClass) with the same name as the previous one. It also responsible for annotating the old machineClass CR with an migrated annotation.
```

Comment:
This reason I have been avoiding this is because,  then [this code](https://github.com/prashanth26/machine-controller-manager-provider-aws/commit/3598508cc0a01983a6baac65322c6bdf6d67b774#diff-28b185322182fd20dde9a7dc05dcb6b4R379) of annotating the old machine class would go into the current reconciliation loop. And this would mean multiple different calls might have to be made to different machineClass clients while annotation, however keeping it inside the provider spec it would be easier to determine which call to exactly make. However, now what I re-think about it. It seems fine to me.

Let me try making such changes with the new GCP provider and test it out. And see if I face any difficulties.

Reply:
👍

The existing in-tree code already has the logic based on the old typed machine classes. So, this would be one more place. Also, this would be only for migration, we can phase this code out via deprecation over a couple of releases. But it if is there in the providers, then phasing it out also is then multiplied by the number of providers.

Reply:
Yes makes sense.

Reply:
I have moved the common logic into this [pkg with this commit](https://github.com/gardener/machine-controller-manager/pull/484/commits/e0fc656b6800e0926087cdb42cc3cb139598f1dd). Also the leftover at the provider can be seen [here at the provider](https://github.com/gardener/machine-controller-manager-provider-gcp/pull/5/commits/34e931665d937c3efa057c594e1d43bf718a33df). I hope this is what you meant.